### PR TITLE
Support for Python 3

### LINKED ISSUE #2: Support Python 3
and prove it by running tests on Python 3.

### PR DESCRIPTION
**Pull request by vernans**
Fixes #2

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -2,6 +2,12 @@ language: python
 sudo: false
 python:
     - 2.7
+    - 3.5
+    - 3.6
+matrix:
+    include:
+        - python: "3.7"
+          dist: xenial
```
Comment (mgedmin) [inline .travis.yml]: Travis CI is fun and exciting: the best practices change all the time!

This month's wisdom is:

- drop `sudo: false`
- move `dist: xenial` to the top level
- list all supported Python versions in the top-level `python:`
- (if you support pypy, you can (again) use `pypy` and `pypy3` to get the latest releases of pypy2.7 and pypy3.5)

File: setup.py [hunk 2]
```diff
@@ -1,51 +1,62 @@
...
-       'Topic :: Software Development :: Build Tools',
-       'Topic :: Software Development :: Libraries :: Python Modules',
-       ],
-    )
+    packages=find_packages('src'),
+    include_package_data=True,
+    package_dir={'': 'src'},
+    namespace_packages=['zc', 'zc.recipe'],
+    extras_require={
+        'tests': ['zope.testing']
```
Comment (mgedmin) [inline setup.py]: I would add a trailing comma after the `]` here.

File: setup.py [hunk 3]
```diff
@@ -1,51 +1,62 @@
...
+    },
+    install_requires=[
+        'zc.buildout >=1.2.0',
+        'zope.testrunner',
+        'setuptools',
+        'zc.recipe.egg  >=1.2.0',
```
Comment (mgedmin) [inline setup.py]: While we're fixing indentation maybe let's normalize the varying amounts of spaces in front of `>=` in these version pins?

File: setup.py [hunk 4]
```diff
@@ -1,51 +1,62 @@
...
+    ],
+    tests_require=['zope.testing'],
+    test_suite=name + '.tests.test_suite',
+    entry_points={
+        'zc.buildout': ['default = %s:TestRunner' % name]
```
Comment (mgedmin) [inline setup.py]: I would add a trailing comma after the `]`.

File: src/zc/recipe/testrunner/README.txt [hunk 5]
```diff
@@ -387,7 +387,7 @@ Running them with the current buildout will produce a failure:
     Running zope.testrunner.layer.UnitTests tests:
       Set up zope.testrunner.layer.UnitTests in 0.001 seconds.
       Running:
-     test (demo.tests.DemoTests) (0.000 s)
+     test (demo.tests.DemoTests) (... s)
```
Comment (mgedmin) [inline src/zc/recipe/testrunner/README.txt]: Huh, there's a renormalizer for `\d[.]\d+ seconds` to `0.001 seconds`, but no renormalizer for the `s` variant:

https://github.com/zopefoundation/zc.recipe.testrunner/blob/766796a13381d21a2b4bc24499ca93aae2d7e81a/src/zc/recipe/testrunner/tests.py#L45

I'd add a second renormalizer instead of using `...` in the doctest.

File: src/zc/recipe/testrunner/bugfixes.txt [hunk 6]
```diff
@@ -1,104 +0,0 @@
...
-Wrong setting of working-dir
-============================
-
-Since version 3.6.0 zope.testing can run layers in parallel
-processes. zc.recipe.testrunner sets the working directory for each
-using a relative path. The subprocesses have the given working
-directory already set, but zc.recipe.testrunner tried to set it again
-and failed ("IOError: No such file or directory").
-
-To illustrate this, we'll create a project with two layers:
```
Comment (mgedmin) [inline src/zc/recipe/testrunner/bugfixes.txt]: Why did you remove this regression test?

File: src/zc/recipe/testrunner/tests.py [hunk 7]
```diff
@@ -12,57 +12,166 @@
 #
 ##############################################################################
 
+from zc.buildout.testing import mkdir, write, system
+
```
Comment (mgedmin) [inline src/zc/recipe/testrunner/tests.py]: PEP-8 says

- stdlib imports first
- 3rd-party libraries next (zc.buildout would go there)
- local imports last

File: src/zc/recipe/testrunner/bugfixes.txt (hunk 6 above)
Comment (mgedmin) [reply src/zc/recipe/testrunner/bugfixes.txt]: Ah, I see you moved it into `tests.py`.

File: src/zc/recipe/testrunner/tests.py [hunk 8]
```diff
@@ -12,57 +12,166 @@
...
 import unittest
-
 import zc.buildout.testing
-import doctest
 import zope.testing.renormalizing
 
 
+SETUP_PY = """from setuptools import setup
+setup(name = "bugfix1")
+"""
```
Comment (mgedmin) [inline src/zc/recipe/testrunner/tests.py]: Usually multiline strings are more readable if you insert a line break after the opening `"""`:

```python
SETUP_PY = """\
from setuptools import setup
setup(name="bugfix1")
"""
```

If you strongly dislike backslashes, another option is to `.lstrip('\n')`:


```python
SETUP_PY = """
from setuptools import setup
setup(name="bugfix1")
""".lstrip('\n')
```

When the multiline string literals are very long, or are defined in some nested scope, they can be made more readable with the help of textwrap.dedent:

```python
SETUP_PY = textwrap.dedent("""\
    from setuptools import setup
    setup(name="bugfix1")
""")
```

I do not think this is necessary here and am merely mentioning it for completeness.

File: src/zc/recipe/testrunner/tests.py [hunk 9]
```diff
@@ -12,57 +12,166 @@
...
+Total: 2 tests, 0 failures, 0 errors and 0 skipped in
+"""
+
+
+class AbsPathTest(unittest.TestCase):
+    """Since version 3.6.0 zope.testing can run layers in parallel
+    processes. zc.recipe.testrunner sets the working directory for each
+    using a relative path. The subprocesses have the given working
+    directory already set, but zc.recipe.testrunner tried to set it again
+    and failed ("IOError: No such file or directory")."""
```
Comment (mgedmin) [inline src/zc/recipe/testrunner/tests.py]: PEP-257 recommends moving the closing `"""` to a line of its own, for multi-line docstrings.

File: src/zc/recipe/testrunner/tests.py [hunk 10]
```diff
@@ -12,57 +12,166 @@
...
+
+    def runTest(self):
+        output = system(os.path.join(self.tmp, 'bin', 'test') + ' -vv -j2')
+        comp_lines = OUTPUT_COMP.split('\n')
+
+        # Here we check if meaningful outputs have been given. See above.
+        self.assertTrue(OUTPUT_COMP)
+        self.assertTrue(comp_lines)
+        for line in comp_lines:
+            self.assertIn(line, output)
```
Comment (mgedmin) [inline src/zc/recipe/testrunner/tests.py]: Python 2.7+ has this nifty [self.assertMultiLineEqual](https://docs.python.org/2/library/unittest.html#unittest.TestCase.assertMultiLineEqual).

It will be unsuitable if the intent here is to allow additional lines in the output (e.g. setuptools warnings etc) that should not break the test by their presence.

File: src/zc/recipe/testrunner/tests.py [hunk 11]
```diff
@@ -12,57 +12,166 @@
...
-                     #more windows happiness:
-                     (re.compile(r'eggs\\\\'),
-                                'eggs/'),
-                     (re.compile(r'parts\\\\'),
-                                'parts/'),
-                     ]),
-            ),
+                [zc.buildout.testing.normalize_path,
+                    zc.buildout.testing.normalize_script,
+                    zc.buildout.testing.normalize_egg_py,
```
Comment (mgedmin) [inline src/zc/recipe/testrunner/tests.py]: The indentation here is weirdly inconsistent.  I suggest

```python
                [
                    zc.buildout.testing.normalize_path,
                    zc.buildout.testing.normalize_script,
                    zc.buildout.testing.normalize_egg_py,￼
```

File: src/zc/recipe/testrunner/tests.py [hunk 12]
```diff
@@ -12,57 +12,166 @@
...
+                    zc.buildout.testing.normalize_endings,
+                    (re.compile('#!\S+py\S*'), '#!python'),
+                    (re.compile('\d[.]\d+ seconds'), '0.001 seconds'),
```
Comment (mgedmin) [inline src/zc/recipe/testrunner/tests.py]: And here I suggest adding a

```python
                    (re.compile(r'\d[.]\d+ s\b'), '0.000 s'),
```

instead of relying on the `...` in the doctest.

File: src/zc/recipe/testrunner/tests.py [hunk 13]
```diff
@@ -12,57 +12,166 @@
...
-                                'eggs/'),
-                     (re.compile(r'parts\\\\'),
-                                'parts/'),
-                     ]),
-            ),
+                [zc.buildout.testing.normalize_path,
+                    zc.buildout.testing.normalize_script,
+                    zc.buildout.testing.normalize_egg_py,
+                    zc.buildout.testing.normalize_endings,
+                    (re.compile('#!\S+py\S*'), '#!python'),
```
Comment (mgedmin) [inline src/zc/recipe/testrunner/tests.py]: BTW Python 3.8 started emitting SyntaxWarnings for all unknown escape sequences, so it'll be time to switch to using raw strings for these soon.

(Maybe not in this PR: scope creep is a thing.)

File: src/zc/recipe/testrunner/tests.py (hunk 10 above)
Comment (vernans) [reply src/zc/recipe/testrunner/tests.py]: Yes, this was the intent. But I havent't known assertMultiLineEqual, thanks!

**Comment by vernans** [comment]
@mgedmin Thanks for the review!
